File: app/src/main/java/com/omar/retromp3recorder/app/player/AudioPlayer.java
package com.omar.retromp3recorder.app.player;

import io.reactivex.Observable;

public interface AudioPlayer {

    void playerStop();

    void playerStart(String voiceURL);

    Observable<Event> observeEvents();

    //region events
    interface Event { }

    final class Message implements Event {
        public final String message;

        public Message(String errorMessage) {
            this.message = errorMessage;
        }
    }

    final class Error implements Event {
        public final String error;

        public Error(String error) {
            this.error = error;
        }
    }

    final class SendPlayerId implements Event {
        final int playerId;

        SendPlayerId(int playerId) {
            this.playerId = playerId;
        }
    }

    final class PlaybackEnded implements Event { }
    //endregion
}

File: app/src/main/java/com/omar/retromp3recorder/app/playback/player/AudioPlayer.kt
package com.omar.retromp3recorder.app.playback.player

import com.github.alkurop.stringerbell.Stringer
import io.reactivex.Observable

interface AudioPlayer {
    fun playerStop()
    fun playerStart(voiceURL: String)
    fun observeEvents(): Observable<Event>
    val isPlaying: Boolean

    //region events
    sealed class Event {
        data class Message(val message: Stringer) : Event()
        data class Error(val error: Stringer) : Event()
        data class PlayerId internal constructor(val playerId: Int) : Event()
    }
}